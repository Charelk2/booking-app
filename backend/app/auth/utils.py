import re
import secrets
from typing import Optional

from starlette.responses import Response

from app.core.config import COOKIE_DOMAIN

SAFE_NEXT_RE = re.compile(r"^/[a-zA-Z0-9\-._~/]*(\?[a-zA-Z0-9\-._~&=%]*)?$")


def new_oauth_state() -> str:
    """Return a cryptographically secure OAuth state token."""
    return secrets.token_urlsafe(24)


def sanitize_next(next_path: Optional[str]) -> str:
    """Limit redirects to safe, relative in-app paths."""
    if not next_path:
        return "/"
    if not isinstance(next_path, str):
        return "/"
    candidate = next_path.strip()
    if not candidate:
        return "/"
    if not SAFE_NEXT_RE.match(candidate):
        return "/"
    return candidate


def set_session_cookie(response: Response, name: str, value: str, max_age: int = 30 * 24 * 60 * 60) -> None:
    """Set a cross-site-friendly session cookie on the callback response."""
    if not name:
        raise ValueError("Cookie name is required")
    cookie_domain = COOKIE_DOMAIN or None
    response.set_cookie(
        key=name,
        value=value,
        domain=cookie_domain,
        path="/",
        secure=True,
        httponly=True,
        samesite="none",
        max_age=max_age,
    )


__all__ = ["new_oauth_state", "sanitize_next", "set_session_cookie", "SAFE_NEXT_RE"]
