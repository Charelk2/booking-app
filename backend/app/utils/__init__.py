from .json_utils import dumps
from .messages import parse_booking_details
from .errors import error_response
