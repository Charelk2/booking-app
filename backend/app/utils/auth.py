from passlib.context import CryptContext

pwd_context = CryptContext(schemes=["bcrypt"], deprecated="auto")


def get_password_hash(password: str) -> str:
    return pwd_context.hash(password)


def verify_password(plain_password: str, hashed_password: str) -> bool:
    return pwd_context.verify(plain_password, hashed_password)


def normalize_email(email: str) -> str:
    """Return a normalized email address for comparison and storage.

    Gmail addresses are canonicalized so aliases like "user+tag@googlemail.com"
    and "u.s.e.r@gmail.com" resolve to the same account. This prevents duplicate
    users when logging in via Google OAuth.
    """

    email = email.strip().lower()
    local, _, domain = email.partition("@")

    if domain in {"gmail.com", "googlemail.com"}:
        domain = "gmail.com"
        local = local.split("+", 1)[0].replace(".", "")

    return f"{local}@{domain}"
