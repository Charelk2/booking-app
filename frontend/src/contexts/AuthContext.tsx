'use client';

import { createContext, useContext, useState, useEffect, ReactNode } from 'react';
import { User } from '@/types';
import {
  login as apiLogin,
  register as apiRegister,
  verifyMfa as apiVerifyMfa,
  confirmMfa as apiConfirmMfa,
  generateRecoveryCodes as apiGenerateRecoveryCodes,
  disableMfa as apiDisableMfa,
} from '@/lib/api';

interface AuthContextType {
  user: User | null;
  token: string | null;
  loading: boolean;
  login: (
    email: string,
    password: string,
    remember?: boolean,
  ) => Promise<{ mfaRequired: true; token: string } | void>;
  verifyMfa: (
    token: string,
    code: string,
    remember?: boolean,
  ) => Promise<void>;
  confirmMfa: (code: string) => Promise<void>;
  generateRecoveryCodes: () => Promise<string[]>;
  disableMfa: (code: string) => Promise<void>;
  register: (data: Partial<User>) => Promise<void>;
  logout: () => void;
}

const AuthContext = createContext<AuthContextType | undefined>(undefined);

export function AuthProvider({ children }: { children: ReactNode }) {
  const [user, setUser] = useState<User | null>(null);
  const [token, setToken] = useState<string | null>(null);
  const [loading, setLoading] = useState(true);

  useEffect(() => {
    const storedUser =
      localStorage.getItem('user') || sessionStorage.getItem('user');
    const storedToken =
      localStorage.getItem('token') || sessionStorage.getItem('token');
    if (storedUser) {
      try {
        setUser(JSON.parse(storedUser));
      } catch (e) {
        console.error('Failed to parse stored user:', e);
        localStorage.removeItem('user');
        sessionStorage.removeItem('user');
      }
    }
    if (storedToken) {
      setToken(storedToken);
    }
    setLoading(false);
  }, []);

  const login = async (
    email: string,
    password: string,
    remember = false,
  ) => {
    try {
      const response = await apiLogin(email, password);
      if (response.data.mfa_required) {
        return { mfaRequired: true, token: response.data.mfa_token } as const;
      }
      const { user: userData, access_token } = response.data;
      setUser(userData);
      setToken(access_token);
      const storage = remember ? localStorage : sessionStorage;
      const altStorage = remember ? sessionStorage : localStorage;
      storage.setItem('user', JSON.stringify(userData));
      storage.setItem('token', access_token);
      altStorage.removeItem('user');
      altStorage.removeItem('token');
    } catch (error) {
      console.error('Login failed:', error);
      throw error;
    }
  };

  const verifyMfa = async (
    tokenToVerify: string,
    code: string,
    remember = false,
  ) => {
    try {
      const response = await apiVerifyMfa(tokenToVerify, code);
      const { user: userData, access_token } = response.data;
      setUser(userData);
      setToken(access_token);
      const storage = remember ? localStorage : sessionStorage;
      const altStorage = remember ? sessionStorage : localStorage;
      storage.setItem('user', JSON.stringify(userData));
      storage.setItem('token', access_token);
      altStorage.removeItem('user');
      altStorage.removeItem('token');
    } catch (error) {
      console.error('MFA verification failed:', error);
      throw error;
    }
  };

  const confirmMfa = async (code: string) => {
    await apiConfirmMfa(code);
    setUser((prev) => (prev ? { ...prev, mfa_enabled: true } : prev));
  };

  const generateRecoveryCodes = async () => {
    const res = await apiGenerateRecoveryCodes();
    return res.data.codes as string[];
  };

  const disableMfa = async (code: string) => {
    await apiDisableMfa(code);
    setUser((prev) => (prev ? { ...prev, mfa_enabled: false } : prev));
  };

  const register = async (data: Partial<User>) => {
    try {
      const response = await apiRegister(data);
      const userData = response.data;
      setUser(userData);
      localStorage.setItem('user', JSON.stringify(userData));
    } catch (error) {
      console.error('Registration failed:', error);
      throw error;
    }
  };

  const logout = () => {
    setUser(null);
    setToken(null);
    localStorage.removeItem('user');
    localStorage.removeItem('token');
    sessionStorage.removeItem('user');
    sessionStorage.removeItem('token');
  };

  return (
    <AuthContext.Provider
      value={{
        user,
        token,
        loading,
        login,
        verifyMfa,
        confirmMfa,
        generateRecoveryCodes,
        disableMfa,
        register,
        logout,
      }}
    >
      {children}
    </AuthContext.Provider>
  );
}

export function useAuth() {
  const context = useContext(AuthContext);
  if (context === undefined) {
    throw new Error('useAuth must be used within an AuthProvider');
  }
  return context;
} 