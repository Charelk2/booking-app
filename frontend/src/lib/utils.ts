import api from './api';

export const getFullImageUrl = (relativePath: string | undefined | null): string | null => {
  if (!relativePath) return null;
  if (relativePath.startsWith('http://') || relativePath.startsWith('https://')) {
    return relativePath;
  }
  const cleanPath = relativePath.startsWith('/static/') ? relativePath : `/static/${relativePath.replace(/^\/+/, '')}`;
  return `${api.defaults.baseURL}${cleanPath}`;
};

export const extractErrorMessage = (detail: unknown): string => {
  if (!detail) return 'An unexpected error occurred.';
  if (typeof detail === 'string') return detail;
  if (Array.isArray(detail)) {
    return detail
      .map((d) => {
        if (typeof d === 'object' && d && 'msg' in d) {
          const record = d as Record<string, unknown>;
          return record.msg as string;
        }
        return JSON.stringify(d);
      })
      .join(', ');
  }
  if (typeof detail === 'object') {
    const record = detail as Record<string, unknown>;
    if (typeof record.msg === 'string') return record.msg;
    return JSON.stringify(detail);
  }
  return String(detail);
};
