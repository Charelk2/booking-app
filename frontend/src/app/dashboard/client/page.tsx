"use client";

import React, { useEffect, useState, useMemo } from "react";
import { useRouter, usePathname, useSearchParams } from "next/navigation";
import MainLayout from "@/components/layout/MainLayout";
import { useAuth } from "@/contexts/AuthContext";
import { Booking, BookingRequest } from "@/types";
import { getMyClientBookings, getMyBookingRequests } from "@/lib/api";
import SectionList from "@/components/dashboard/SectionList";
import BookingRequestCard from "@/components/dashboard/BookingRequestCard";
import { Spinner } from "@/components/ui";
import DashboardTabs from "@/components/dashboard/DashboardTabs";
import { format } from "date-fns";
import { formatCurrency, formatStatus } from "@/lib/utils";
import Link from "next/link";

export default function ClientDashboardPage() {
  const { user, loading: authLoading } = useAuth();
  const router = useRouter();
  const pathname = usePathname();
  const params = useSearchParams();

  const [bookings, setBookings] = useState<Booking[]>([]);
  const [bookingRequests, setBookingRequests] = useState<BookingRequest[]>([]);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState("");

  const initialTab = params.get("tab") === "bookings" ? "bookings" : "requests";
  const [activeTab, setActiveTab] = useState<"requests" | "bookings">(initialTab);

  useEffect(() => {
    const params = new URLSearchParams(window.location.search);
    params.set("tab", activeTab);
    router.replace(`${pathname}?${params.toString()}`);
  }, [activeTab, router, pathname]);

  useEffect(() => {
    if (authLoading) return;
    if (!user) {
      router.push(`/login?next=${encodeURIComponent(pathname)}`);
      return;
    }
    if (user.user_type !== "client") {
      router.push("/dashboard/artist");
      return;
    }

    const fetchData = async () => {
      try {
        const [bookingsData, requestsData] = await Promise.all([
          getMyClientBookings(),
          getMyBookingRequests(),
        ]);
        setBookings(bookingsData.data);
        setBookingRequests(requestsData.data);
      } catch (err) {
        console.error("Client dashboard fetch error:", err);
        setError("Failed to load dashboard data. Please try again.");
      } finally {
        setLoading(false);
      }
    };

    fetchData();
  }, [user, authLoading, router, pathname]);

  const upcomingBookings = useMemo(() => {
    const now = new Date().getTime();
    return bookings
      .filter((b) => new Date(b.start_time).getTime() >= now)
      .sort((a, b) => new Date(a.start_time).getTime() - new Date(b.start_time).getTime())
      .slice(0, 5);
  }, [bookings]);

  if (loading) {
    return (
      <MainLayout>
        <div className="p-8 flex justify-center"><Spinner /></div>
      </MainLayout>
    );
  }

  if (error) {
    return (
      <MainLayout>
        <div className="p-8 text-red-600">{error}</div>
      </MainLayout>
    );
  }

  return (
    <MainLayout>
      <div className="max-w-5xl mx-auto px-4 py-8">
        <DashboardTabs activeTab={activeTab} onChange={setActiveTab} showServices={false} />
        <div className="mt-6">
          {activeTab === "requests" && (
            <section>
              <SectionList
                title="Booking Requests"
                data={bookingRequests}
                renderItem={(r) => <BookingRequestCard key={r.id} request={r} onUpdate={() => {}} />}
                emptyState={<span>No requests yet</span>}
              />
            </section>
          )}
          {activeTab === "bookings" && (
            <section>
              <SectionList
                title="Upcoming Bookings"
                data={upcomingBookings}
                emptyState={<span>No bookings yet</span>}
                renderItem={(booking) => (
                  <div key={booking.id} className="bg-white p-4 shadow rounded-lg">
                    <div className="font-medium text-gray-900">
                      {booking.artist?.first_name || "Unknown"} {booking.artist?.last_name || ""}
                    </div>
                    <div className="text-sm text-gray-500">{booking.service?.title || "â€”"}</div>
                    <div className="text-sm text-gray-500">
                      {format(new Date(booking.start_time), "MMM d, yyyy h:mm a")}
                    </div>
                    <div className="mt-2 flex justify-between items-center">
                      <span
                        className={`inline-flex rounded-full px-2 text-xs font-semibold leading-5 ${
                          booking.status === "completed"
                            ? "bg-green-100 text-green-800"
                            : booking.status === "cancelled"
                            ? "bg-red-100 text-red-800"
                            : booking.status === "confirmed"
                            ? "bg-brand-light text-brand-dark"
                            : "bg-yellow-100 text-yellow-800"
                        }`}
                      >
                        {formatStatus(booking.status)}
                      </span>
                      <span className="text-sm text-gray-500">
                        {formatCurrency(Number(booking.total_price))}
                      </span>
                    </div>
                  </div>
                )}
              />
              {bookings.length > upcomingBookings.length && (
                <div className="mt-2">
                  <Link href="/dashboard/client/bookings" className="text-brand-dark hover:underline text-sm">
                    View All Bookings
                  </Link>
                </div>
              )}
            </section>
          )}
        </div>
      </div>
    </MainLayout>
  );
}

