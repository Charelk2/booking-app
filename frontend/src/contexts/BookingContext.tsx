import { createContext, useContext, useState, ReactNode, useEffect } from 'react';
import { TravelResult } from '@/lib/travel';

export interface EventDetails {
  eventType: string;
  eventDescription: string;
  date: Date;
  /** Optional time string in HH:mm format */
  time?: string;
  location: string;
  guests: string;
  venueType: 'indoor' | 'outdoor' | 'hybrid';
  sound: 'yes' | 'no';
  notes?: string;
  attachment_url?: string;
}

interface BookingContextValue {
  step: number;
  setStep: (s: number) => void;
  details: EventDetails;
  setDetails: (d: EventDetails) => void;
  serviceId?: number;
  setServiceId: (id: number | undefined) => void;
  requestId?: number;
  setRequestId: (id: number | undefined) => void;
  travelResult: TravelResult | null;
  setTravelResult: (r: TravelResult | null) => void;
  resetBooking: () => void;
  loadSavedProgress: () => boolean;
}

const BookingContext = createContext<BookingContextValue | undefined>(undefined);

const STORAGE_KEY = 'bookingState';

const initialDetails: EventDetails = {
  eventType: '',
  eventDescription: '',
  date: new Date(),
  time: '',
  location: '',
  guests: '',
  venueType: 'indoor',
  sound: 'yes',
  attachment_url: '',
};

export const BookingProvider = ({ children }: { children: ReactNode }) => {
  const [step, setStep] = useState(0);
  const [details, setDetails] = useState<EventDetails>(initialDetails);
  const [serviceId, setServiceId] = useState<number | undefined>(undefined);
  const [requestId, setRequestId] = useState<number | undefined>(undefined);
  const [travelResult, setTravelResult] = useState<TravelResult | null>(null);

  const resetBooking = () => {
    setStep(0);
    setDetails(initialDetails);
    setServiceId(undefined);
    setRequestId(undefined);
    setTravelResult(null);
    if (typeof window !== 'undefined') {
      localStorage.removeItem(STORAGE_KEY);
    }
  };

  /**
   * Restore an in-progress booking from localStorage if the user confirms.
   * Returns true when progress was restored.
   */
  const loadSavedProgress = () => {
    if (typeof window === 'undefined') return false;
    const stored = localStorage.getItem(STORAGE_KEY);
    if (!stored) return false;
    try {
      const parsed = JSON.parse(stored) as {
        step?: number;
        details?: Partial<EventDetails> & { date?: string; time?: string };
        serviceId?: number;
        requestId?: number;
        travelResult?: TravelResult | null;
      };

      const resume = window.confirm(
        'Resume your previous booking request? Choose Cancel to start over.',
      );
      if (!resume) {
        localStorage.removeItem(STORAGE_KEY);
        return false;
      }

      if (parsed.step !== undefined) setStep(parsed.step);
      if (parsed.details) {
        const parsedDetails: EventDetails = {
          ...initialDetails,
          ...parsed.details,
          date: parsed.details.date ? new Date(parsed.details.date) : new Date(),
        };
        setDetails(parsedDetails);
      }
      if (parsed.serviceId !== undefined) setServiceId(parsed.serviceId);
      if (parsed.requestId !== undefined) setRequestId(parsed.requestId);
      if (parsed.travelResult) setTravelResult(parsed.travelResult);

      return true;
    } catch (e) {
      console.error('Failed to parse saved booking progress:', e);
      localStorage.removeItem(STORAGE_KEY);
      return false;
    }
  };

  // Persist progress
  useEffect(() => {
    if (typeof window === 'undefined') return;
    const data = {
      step,
        details: {
          ...details,
          // Guard against undefined or invalid dates when persisting progress
          date: details.date ? new Date(details.date).toISOString() : null,
        },
      serviceId,
      requestId,
      travelResult,
    };
    try {
      localStorage.setItem(STORAGE_KEY, JSON.stringify(data));
    } catch (e) {
      console.error('Failed to save booking progress:', e);
    }
  }, [step, details, serviceId, requestId, travelResult]);
  return (
    <BookingContext.Provider
      value={{
        step,
        setStep,
        details,
        setDetails,
        serviceId,
        setServiceId,
      requestId,
      setRequestId,
      travelResult,
      setTravelResult,
      resetBooking,
      loadSavedProgress,
    }}
    >
      {children}
    </BookingContext.Provider>
  );
};

export const useBooking = () => {
  const ctx = useContext(BookingContext);
  if (!ctx) throw new Error('useBooking must be used within BookingProvider');
  return ctx;
};
