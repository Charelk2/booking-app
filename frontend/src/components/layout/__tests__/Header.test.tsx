import { flushPromises } from "@/test/utils/flush";
import { createRoot } from 'react-dom/client';
import React from 'react';
import { act } from 'react';
import Header from '../Header';
import { usePathname } from 'next/navigation';
import { useAuth } from '@/contexts/AuthContext';

jest.mock('next/link', () => ({ __esModule: true, default: (props: Record<string, unknown>) => <a {...props} /> }));
jest.mock('next/navigation', () => ({
  usePathname: jest.fn(),
  useRouter: () => ({ push: jest.fn() }),
  useSearchParams: () => new URLSearchParams(),
}));
jest.mock('@/contexts/AuthContext', () => ({ useAuth: jest.fn(() => ({ user: null, logout: jest.fn() })) }));


function render() {
  const div = document.createElement('div');
  document.body.appendChild(div);
  const root = createRoot(div);
  return { div, root };
}

describe('Header', () => {
  afterEach(() => {
    jest.clearAllMocks();
    document.body.innerHTML = '';
  });

  it('renders search bar on home page', async () => {
    (usePathname as jest.Mock).mockReturnValue('/');
    const { div, root } = render();
    await act(async () => {
      root.render(<Header />);
    });
    await flushPromises();
    expect(div.firstChild).toMatchSnapshot();
    act(() => root.unmount());
    div.remove();
  });

  it('renders artists header when extraBar provided', async () => {
    (usePathname as jest.Mock).mockReturnValue('/artists');
    const { div, root } = render();
    await act(async () => {
      root.render(<Header extraBar={<div>bar</div>} />);
    });
    await flushPromises();
    expect(div.firstChild).toMatchSnapshot();
    act(() => root.unmount());
    div.remove();
  });
});
