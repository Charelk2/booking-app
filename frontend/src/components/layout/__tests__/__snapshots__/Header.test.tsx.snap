// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header hides search bar on other pages 1`] = `
<header
  class="sticky top-0 z-40 bg-gray-50"
>
  <div
    class="mx-auto max-w-7xl px-4 sm:px-6 lg:px-8"
  >
    <div
      class="h-16 grid grid-cols-[auto,1fr,auto] items-center"
    >
      <div
        class="flex flex-col"
      >
        <a
          class="flex shrink-0 items-center text-xl font-bold text-brand-dark"
          href="/"
        >
          Booka.co.za
        </a>
      </div>
      <nav
        class="hidden md:flex justify-center gap-6"
      >
        <a
          class="inline-flex items-center border-b-2 px-1 pt-1 text-sm font-medium transition-colors border-primary text-gray-900"
          href="/artists"
        >
          Artists
        </a>
        <a
          class="inline-flex items-center border-b-2 px-1 pt-1 text-sm font-medium transition-colors border-transparent text-gray-500 hover:text-gray-700 hover:border-gray-300"
          href="/services"
        >
          Services
        </a>
        <a
          class="inline-flex items-center border-b-2 px-1 pt-1 text-sm font-medium transition-colors border-transparent text-gray-500 hover:text-gray-700 hover:border-gray-300"
          href="/faq"
        >
          FAQ
        </a>
        <a
          class="inline-flex items-center border-b-2 px-1 pt-1 text-sm font-medium transition-colors border-transparent text-gray-500 hover:text-gray-700 hover:border-gray-300"
          href="/contact"
        >
          Contact
        </a>
      </nav>
      <div
        class="hidden sm:flex items-center justify-end space-x-4"
      >
        <div
          class="space-x-4"
        >
          <a
            class="text-gray-500 hover:text-gray-700 px-3 py-2 text-sm font-medium"
            href="/login"
          >
            Sign in
          </a>
          <a
            class="bg-brand-dark text-white hover:bg-brand-dark px-3 py-2 rounded-md text-sm font-medium"
            href="/register"
          >
            Sign up
          </a>
        </div>
      </div>
    </div>
    <div
      class="flex items-center sm:hidden"
    >
      <button
        class="-mr-2 ml-2 inline-flex items-center justify-center rounded-md p-2 text-gray-400 hover:bg-gray-100 hover:text-gray-500 focus:outline-none focus:ring-2 focus:ring-inset focus:ring-brand"
      >
        <span
          class="sr-only"
        >
          Open main menu
        </span>
        <svg
          aria-hidden="true"
          class="h-6 w-6"
          data-slot="icon"
          fill="none"
          stroke="currentColor"
          stroke-width="1.5"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M3.75 6.75h16.5M3.75 12h16.5m-16.5 5.25h16.5"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
      </button>
    </div>
  </div>
</header>
`;

exports[`Header renders search bar on home page 1`] = `
<header
  class="sticky top-0 z-40 bg-gray-50"
>
  <div
    class="mx-auto max-w-7xl px-4 sm:px-6 lg:px-8"
  >
    <div
      class="h-16 grid grid-cols-[auto,1fr,auto] items-center"
    >
      <div
        class="flex flex-col"
      >
        <a
          class="flex shrink-0 items-center text-xl font-bold text-brand-dark"
          href="/"
        >
          Booka.co.za
        </a>
        <span
          class="hidden sm:block -mt-0.5 text-[11px] leading-4 text-gray-600"
        >
          Book legendary artists
        </span>
      </div>
      <nav
        class="hidden md:flex justify-center gap-6"
      >
        <a
          class="inline-flex items-center border-b-2 px-1 pt-1 text-sm font-medium transition-colors border-transparent text-gray-500 hover:text-gray-700 hover:border-gray-300"
          href="/artists"
        >
          Artists
        </a>
        <a
          class="inline-flex items-center border-b-2 px-1 pt-1 text-sm font-medium transition-colors border-transparent text-gray-500 hover:text-gray-700 hover:border-gray-300"
          href="/services"
        >
          Services
        </a>
        <a
          class="inline-flex items-center border-b-2 px-1 pt-1 text-sm font-medium transition-colors border-transparent text-gray-500 hover:text-gray-700 hover:border-gray-300"
          href="/faq"
        >
          FAQ
        </a>
        <a
          class="inline-flex items-center border-b-2 px-1 pt-1 text-sm font-medium transition-colors border-transparent text-gray-500 hover:text-gray-700 hover:border-gray-300"
          href="/contact"
        >
          Contact
        </a>
      </nav>
      <div
        class="hidden sm:flex items-center justify-end space-x-4"
      >
        <div
          class="space-x-4"
        >
          <a
            class="text-gray-500 hover:text-gray-700 px-3 py-2 text-sm font-medium"
            href="/login"
          >
            Sign in
          </a>
          <a
            class="bg-brand-dark text-white hover:bg-brand-dark px-3 py-2 rounded-md text-sm font-medium"
            href="/register"
          >
            Sign up
          </a>
        </div>
      </div>
    </div>
    <div
      class="flex items-center sm:hidden"
    >
      <button
        class="-mr-2 ml-2 inline-flex items-center justify-center rounded-md p-2 text-gray-400 hover:bg-gray-100 hover:text-gray-500 focus:outline-none focus:ring-2 focus:ring-inset focus:ring-brand"
      >
        <span
          class="sr-only"
        >
          Open main menu
        </span>
        <svg
          aria-hidden="true"
          class="h-6 w-6"
          data-slot="icon"
          fill="none"
          stroke="currentColor"
          stroke-width="1.5"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M3.75 6.75h16.5M3.75 12h16.5m-16.5 5.25h16.5"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
      </button>
    </div>
    <div
      class="pb-3 pt-2"
    >
      <div
        class="relative"
      >
        <form
          class="flex items-stretch bg-white rounded-full ring-1 ring-gray-200 shadow-md overflow-visible min-h-[48px] text-sm mx-auto w-full md:max-w-2xl"
        >
          <div
            class="flex-1 px-4 flex flex-col text-left py-2"
          >
            <span
              class="text-xs text-gray-500"
            >
              Category
            </span>
            <div
              class="relative w-full"
            >
              <button
                aria-expanded="false"
                aria-haspopup="listbox"
                class="mt-1 w-full flex justify-between items-center text-sm text-gray-700 focus:outline-none"
                data-headlessui-state=""
                id="headlessui-listbox-button-:r0:"
                type="button"
              >
                <span>
                  Musician / Band
                </span>
                <svg
                  aria-hidden="true"
                  class="h-4 w-4 text-gray-400"
                  data-slot="icon"
                  fill="none"
                  stroke="currentColor"
                  stroke-width="1.5"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="m19.5 8.25-7.5 7.5-7.5-7.5"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                  />
                </svg>
              </button>
            </div>
          </div>
          <div
            class="border-l border-gray-200"
          />
          <div
            class="flex-1 px-4 flex flex-col text-left py-2"
          >
            <span
              class="text-xs text-gray-500"
            >
              Where
            </span>
            <div
              class="relative"
            >
              <div
                class="w-full"
              >
                <div
                  class="mt-1 relative"
                >
                  <input
                    class="block w-full rounded-md border-gray-300 shadow-sm focus:border-brand focus:ring-brand sm:text-sm mt-1 w-full text-sm text-gray-700 placeholder-gray-400 focus:outline-none border-none shadow-none rounded-none bg-transparent p-0 focus:ring-0 pr-8"
                    data-testid="location-input"
                    id=":r1:"
                    placeholder="City or venue"
                  />
                </div>
              </div>
              <button
                class="absolute inset-y-0 right-1 flex items-center text-gray-500 hover:text-gray-700"
                data-testid="open-map-modal"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="h-5 w-5"
                  data-slot="icon"
                  fill="none"
                  stroke="currentColor"
                  stroke-width="1.5"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M15 10.5a3 3 0 1 1-6 0 3 3 0 0 1 6 0Z"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                  />
                  <path
                    d="M19.5 10.5c0 7.142-7.5 11.25-7.5 11.25S4.5 17.642 4.5 10.5a7.5 7.5 0 1 1 15 0Z"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                  />
                </svg>
              </button>
            </div>
          </div>
          <div
            class="border-l border-gray-200"
          />
          <div
            class="flex-1 px-4 flex flex-col text-left py-2"
          >
            <span
              class="text-xs text-gray-500"
            >
              When
            </span>
            <input
              class="mt-1 w-full text-sm text-gray-700 focus:outline-none"
              dateformat="MMM d, yyyy h:mm aa"
              placeholdertext="Add date & time"
            />
          </div>
          <button
            class="bg-pink-600 hover:bg-pink-700 text-white flex items-center justify-center h-full aspect-square md:w-12 md:h-auto"
            type="submit"
          >
            <svg
              aria-hidden="true"
              class="w-5 h-5"
              data-slot="icon"
              fill="none"
              stroke="currentColor"
              stroke-width="1.5"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="m21 21-5.197-5.197m0 0A7.5 7.5 0 1 0 5.196 5.196a7.5 7.5 0 0 0 10.607 10.607Z"
                stroke-linecap="round"
                stroke-linejoin="round"
              />
            </svg>
            <span
              class="sr-only"
            >
              Search
            </span>
          </button>
        </form>
        <div
          class="pointer-events-none absolute inset-x-6 -bottom-2 h-2 rounded-full bg-black/10 blur-md opacity-30"
        />
      </div>
    </div>
  </div>
  <div
    class="border-t border-gray-200"
  />
</header>
`;
